pkg/secrets/watch.go

```diff
@@ -0,0 +1,266 @@
+	refCount uint
+	done     bool
+}
+
+func newWatcher(ctx context.Context, logger log.Logger, client kubernetes.Interface, config *KubernetesSecretConfig) (*secretWatcher, error) {
+	watcher := &secretWatcher{
+		refCount: 1,
+		done:     false,
+	}
+
+	if err := watcher.start(ctx, client, config); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case e, ok := <-watcher.w.ResultChan():
+				if ok {
+					watcher.update(logger, e)
+					continue
+				}
+
+				// If the application shutdown, we don't care about cleanup.
+				if ctx.Err() != nil {
+					watcher.mu.Lock()
+					defer watcher.mu.Lock()
+					watcher.s = nil
+					return
+				}
+				// If closed unintentionally (i.e. network issues), try and restart it.
+				for {
+					ok, err := watcher.restart(ctx, client, config)
+					if ok {
+						return
+					}
+					// If an error occurred trying to watch, keep retrying.
+					if err == nil {
+						break
+					}
+					_ = logger.Log("msg", "unable to restart secret watcher", "err", err, "namespace", watcher.s.Namespace, "name", watcher.s.Name)
+				}
+			case <-ctx.Done():
+				// The application shutdown, we don't care about cleaning up.
+				watcher.close()
+				return
+			}
+		}
+	}()
+
+	return watcher, nil
+}
+
+func (w *secretWatcher) update(logger log.Logger, e watch.Event) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	switch e.Type {
+	case watch.Modified, watch.Added:
+		secret := e.Object.(*corev1.Secret)
+		w.s = secret
+	case watch.Deleted:
+		w.s = nil
+	case watch.Bookmark:
+		// Disabled explicitly when creating the watch interface.
+	case watch.Error:
+		//nolint:errcheck
+		logger.Log("msg", "watch error event", "namespace", w.s.Namespace, "name", w.s.Name)
+	}
+}
+
+func (w *secretWatcher) secret(config *KubernetesSecretConfig) Secret {
+	fn := SecretFn(func(_ context.Context) (string, error) {
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		if w.s == nil {
+			return "", fmt.Errorf("secret %s/%s not found", config.Namespace, config.Name)
+		}
+		return getKey(w.s, config.Key)
+	})
+	return &fn
+}
+
+func (w *secretWatcher) start(ctx context.Context, client kubernetes.Interface, config *KubernetesSecretConfig) error {
+	var err error
+	w.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector:       fields.OneTermEqualSelector(metav1.ObjectNameField, config.Name).String(),
+		AllowWatchBookmarks: false,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to watch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+
+	// We could wait for the first watch event, but it doesn't notify us if the resource doesn't exist.
+	w.s, err = client.CoreV1().Secrets(config.Namespace).Get(ctx, config.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("unable to fetch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	return nil
+}
+
+// restart returns true and restarts the watch if there are any open secrets.
```

Comment:
nit: can you clarify the docstring a bit? What does the routine do, and when does it return false vs true?